tensorboard/webapp/widgets/line_chart_v2/lib/scale.ts

```diff
@@ -0,0 +1,196 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {scaleLinear, scaleLog} from 'd3';
+
+import {ScaleType} from './scale_types';
+
+export interface Scale {
+  forward(domain: [number, number], range: [number, number], x: number): number;
+
+  reverse(domain: [number, number], range: [number, number], x: number): number;
+
+  nice(minAndMax: [number, number]): [number, number];
+
+  /**
+   * @param lowAndHigh bounds of the tick
+   * @param sizeGuidance approximate number of the ticks. Depending on the lowAndHigh, it
+   * may return less or more ticks.
+   */
+  ticks(lowAndHigh: [number, number], sizeGuidance: number): number[];
+}
+
+export function createScale(type: ScaleType): Scale {
+  switch (type) {
+    case ScaleType.LINEAR:
+      return new LinearScale();
+    case ScaleType.LOG10:
+      return new Log10Scale();
+    default:
+      throw new RangeError(`ScaleType ${type} not supported.`);
+  }
+}
+
+const PADDING_RATIO = 0.05;
+const MIN_SIGNIFICANT_PADDING = 0.01;
+
+class LinearScale implements Scale {
+  private transform(
+    inputSpace: [number, number],
+    outputSpace: [number, number],
+    x: number
+  ): number {
+    const [inputMin, inputMax] = inputSpace;
+    const inputSpread = inputMax - inputMin;
+    const [outputMin, outputMax] = outputSpace;
+    const outputSpread = outputMax - outputMin;
+
+    if (inputSpread === 0) {
+      return outputMin;
+    }
+
+    return (outputSpread / inputSpread) * (x - inputMin) + outputMin;
+  }
+
+  forward(
+    domain: [number, number],
+    range: [number, number],
+    x: number
+  ): number {
+    return this.transform(domain, range, x);
+  }
+
+  reverse(
+    domain: [number, number],
+    range: [number, number],
+    x: number
+  ): number {
+    return this.transform(range, domain, x);
+  }
+
+  nice(minAndMax: [number, number]): [number, number] {
+    let [min, max] = minAndMax;
+    if (max < min) {
+      throw new Error('Unexpected input: min is larger than max');
+    }
+
+    const scale = scaleLinear();
+    const padding =
+      max === min
+        ? // In case both `min` and `max` are 0, we want some padding.
+          Math.max(min * PADDING_RATIO, MIN_SIGNIFICANT_PADDING)
+        : (max - min + Number.EPSILON) * PADDING_RATIO;
+    const [niceMin, niceMax] = scale
+      .domain([min - padding, max + padding])
+      .nice()
+      .domain();
+    return [niceMin, niceMax];
+  }
+
+  ticks(lowAndHigh: [number, number], sizeGuidance: number): number[] {
+    return scaleLinear().domain(lowAndHigh).ticks(sizeGuidance);
+  }
+}
+
+class Log10Scale implements Scale {
+  private transform(x: number): number {
+    return Math.log10(x > 0 ? x : Number.MIN_VALUE);
```

Comment:
This transform surprises me on two counts. First, it’s discontinuous and
non-monotonic: `transform(0.001) === -3` but `transform(0) === 5e-324`.
(`Number.MIN_VALUE` is the smallest *positive* float.) Second, why do we
want to override the standard behavior of `log` on edge cases? I always
thought that `log(0) = -Inf`, `log(x where x < 0) is NaN` was pretty
sensible.

Reply:
First question: `transform(0.001) = -3` and `transform(0) = Math.log10(Number.MIN_VALUE) ~ -323.3` right?

Reply:
Ah, yes, of course. Sounds good then. Thanks.